File: python/DataStructures/PriorityQueue/PriorityQueue.py
from typing import Tuple, List, TypeVar, Generic

T = TypeVar('T')


class PriorityQueue(Generic[T]):
    __items: List[Tuple[T, int]]

    def __init__(self):
        self.__items = []

    def __repr__(self):
        return "{}".format(self.__items)

    def is_empty(self):
        return len(self.__items) == 0

    def enqueue(self, item: T, priority: int):
        for index, prioritisedItem in enumerate(self.__items):
            if priority > prioritisedItem[1]:
                self.__items.insert(index, (item, priority))
                return
        self.__items.append((item, priority))

    def dequeue(self) -> Tuple[T, int] or None:
        if self.is_empty():
            raise Exception('Queue Empty')

        return self.__items.pop(0)

File: lib/src/dataStructures/queue/PriorityQueue.ts
import LinkedList from "../linkedList/LinkedList";
import { MatchFunction } from "../../types";
import { ObservableVersioned } from "../../common";

export interface PrioritisedItem {
  priority: number;
}

export default class PriorityQueue<
  T extends PrioritisedItem
  > extends ObservableVersioned {
  items: LinkedList<T>;

  constructor() {
    super();
    this.items = new LinkedList();
  }

  getSize(): number {
    return this.items.size();
  }

  toString() {
    return this.items.toString();
  }

  isEmpty() {
    return this.items.length === 0;
  }

  toArray() {
    return this.items.toArray();
  }

  /**
   * Removes
   * @param {function} matchFunction Returns true for the object being removed
   * @returns The removed item
   */
  removeMatch(matchFunction: MatchFunction<T>) {
    this.tickVersion();
    return this.items.removeMatch(matchFunction);
  }

  enqueue(newItem: T) {
    let index: number = 0;
    for (const item of this.items) {
      if (newItem.priority > item.priority) {
        // Insert item at this point and return
        this.items.insert(index, newItem);
        return;
      }

      index += 1;
    }

    // Just push onto the end
    this.items.append(newItem);
    this.tickVersion();
  }

  dequeue() {
    if (this.isEmpty()) {
      throw new Error("Queue Empty");
    }

    this.tickVersion();
    return this.items.remove(0);
  }
}
